superset/security/manager.py

```diff
@@ -987,15 +989,28 @@ def raise_for_access(  # pylint: disable=too-many-arguments,too-many-branches
                 )
 
         if datasource or query_context or viz:
+            extra_jwt = None
             if query_context:
                 datasource = query_context.datasource
+                extra_jwt = query_context.extra_jwt
             elif viz:
                 datasource = viz.datasource
+                extra_jwt = viz.extra_jwt
 
             assert datasource
 
+            ds_allowed_in_dashboard = False
+            if feature_flag_manager.is_feature_enabled("DASHBOARD_RBAC"):
+                dashboard_data_context = dashboard_jwt_manager.parse_jwt(extra_jwt)
+
+                if dashboard_data_context:
+                    ds_allowed_in_dashboard = (
+                        datasource.id in dashboard_data_context.dataset_ids
+                    )
+
             if not (
-                self.can_access_schema(datasource)
+                ds_allowed_in_dashboard
```

Comment:
Could you elaborate more on why you need users without direct access to certain datasets to have access to that dataset through dashboards? Can you just create virtual datasets and grant them read access to the dataset instead?

Reply:
in our org every dashboard might serve different set of users which basically means a separated role per dashboard.
so managing access for every dataset on every dashboard is a cumbersome work.
it will require managing all datasets permissions for every role.
this problem scale when having hundreds of dashboards.

additionally, dataset permission mechanism has a defect , when changing a virtual dataset name then permission name changes as well but if it is already associated to a role you'll need to remove it and then remove and re-associate bc permission is based on the name of the dataset and not by the id

Reply:
> I would be happy to do a follow up PR just to support SPA , no worries there

In the SPA project we have already done the work to remove dashboard-specific bootstrap data so that the apps can be cleanly merged. I think rather than adding new bootstrap properties at this stage, it would be better to use an endpoint. That said, if you're okay with this feature breaking in the very near future, I won't stand in your way. The compatibility discussion is a relatively minor issue.

I think the security concerns have already been discussed somewhat in [SIP-51](https://github.com/apache/superset/issues/10408). Anyone using the feature flag should understand that dashboard owners have the ability to grant data access to others arbitrarily. These effects should be well documented. It's an entirely different security profile for the application.

I am of the opinion that the security manager should make a database query to find the information it needs, rather than using this new JWT pattern. I haven't seen anything that indicates that a JWT is actually required here, though I could be mistaken.

Introducing new authorization patterns specific to a given feature flag will make it more difficult to onboard into the Superset codebase, as well as making it more complicated to change system behavior. The JWT pattern also makes it complicated to revoke or change a user's access, and introduces a new responsibility for the client to correctly manage the JWT.

Reply:
The jwt and it's data indicates that the request indeed coming from the Dashboard and with the nessecery dataset permission.

Otherwise upon chart data request it would be impossible to know if the request does originates from a dashboard or not

Reply:
But you could write a query selecting dashboards the user can access which use the dataset that's being requested, no?

Something akin to:
```
query = query(BaseDatasource).join(Slice).join(Dashboard).filter(BaseDatasource.id == datasource.id)
query = DashboardFilter("id", SQLAInterface(Dashboard, db.session)).apply(query)
```

Reply:
What if you want to publish a dashboard to some users, but a subset of these users cannot have access to a couple of more sensitive charts in the dashboard? If seems by turning on the DASHBOARD_RBAC flag, there is no way to accommodate this case, unless you add some kind of dashboard level metadata to indicate this dashboard should not issue a JWT.

Reply:
In this use case we would probably just have 2 separated Dashboards
Just bc there no sense of providing dashboard to a client and show him only have of charts, that is a very bad UX don't you think ? 💆

I know Air BnB  have this have a dashboard use case 
But for us it just doesnt makes sense 🙅

Reply:
I'd say even your external clients may have people with different access levels. Maybe not now, but it's possible. 

Sometimes you want a single dashboard for all audience because it's easier to navigate and easier to communicate---you can share the same short link to all users and executives don't have to jump between two dashboards just to see the more sensitive data. You may also have two duplicate dashboards with 80% of the content the same, but it'd gradually become difficult to maintain.

Reply:
Maybe we need a "nested dashboard" viz plugin 😁